includes/Classifai/Providers/AWS/AmazonPersonalize.php

```diff
@@ -0,0 +1,740 @@
+	}
+
+	/**
+	 * Get recent items based on given arguments.
+	 *
+	 * These will then be re-ranked by Amazon Personalize.
+	 *
+	 * @param array $attributes The block attributes.
+	 * @return array
+	 */
+	public function get_recent_items( array $attributes ): array {
+		$post_type      = $attributes['contentPostType'];
+		$key_attributes = [
+			'terms'    => $attributes['taxQuery'] ?? [],
+			'excluded' => $attributes['excludeId'] ?? 0,
+		];
+		$transient_key  = 'classifai_actions_' . $post_type . md5( maybe_serialize( $key_attributes ) );
+		$items          = get_transient( $transient_key );
+
+		if ( false !== $items && ! empty( $items ) ) {
+			return $items;
+		}
+
+		$query_args = [
+			'posts_per_page'      => 100,
+			'post_status'         => 'publish',
+			'no_found_rows'       => true,
+			'ignore_sticky_posts' => true,
+			'post_type'           => $post_type,
+		];
+
+		// Exclude item on which we are displaying the block.
+		if ( ! empty( $attributes['excludeId'] ) ) {
+			$query_args['post__not_in'] = [ absint( $attributes['excludeId'] ) ];
+		}
+
+		// Handle taxonomy filters.
+		if ( isset( $attributes['taxQuery'] ) && ! empty( $attributes['taxQuery'] ) ) {
+			foreach ( $attributes['taxQuery'] as $taxonomy => $terms ) {
+				if ( ! empty( $terms ) ) {
+					$query_args['tax_query'][] = [
+						'taxonomy' => $taxonomy,
+						'field'    => 'term_id',
+						'terms'    => $terms,
+					];
+				}
+			}
+
+			if ( isset( $query_args['tax_query'] ) && count( $query_args['tax_query'] ) > 1 ) {
+				$query_args['tax_query']['relation'] = 'AND';
+			}
+		}
+
+		/**
+		 * Filters recommended content post arguments.
+		 *
+		 * @since 1.8.0
+		 * @hook classifai_recommended_content_post_args
+		 *
+		 * @param {array} $query_args Array of query args to get posts
+		 * @param {array} $attributes The block attributes.
+		 *
+		 * @return {array} Array of query args to get posts
+		 */
+		$query_args = apply_filters(
+			'classifai_recommended_content_post_args',
+			$query_args,
+			$attributes
+		);
+
+		$items = [];
+		$query = new \WP_Query( $query_args );
+
+		if ( $query->have_posts() ) {
+			while ( $query->have_posts() ) {
+				$query->the_post();
+
+				$items[] = get_the_ID();
+			}
+		}
+
+		wp_reset_postdata();
+
+		if ( ! empty( $content ) ) {
+			set_transient( $transient_key, $items, 6 * \HOUR_IN_SECONDS );
+		}
+
+		return $items;
+	}
+
+	/**
+	 * Send a tracking event to Amazon Personalize.
+	 *
+	 * @param int   $post_id The post ID to track.
+	 * @param array $args Additional event arguments.
+	 * @return bool|WP_Error
+	 */
+	public function track_event( $post_id, array $args = [] ) {
+		$settings = $this->feature_instance->get_settings( static::ID );
+		$client   = $this->get_client( 'personalize-events' );
+
+		if ( ! $client ) {
+			return new WP_Error( 'client_not_found', esc_html__( 'Client not found.', 'classifai' ) );
+		}
+
+		$id     = uniqid(); // TODO: Is this the best way to generate a session ID?
```

Comment:
This will create a new ID for every event. We'll need to generate a session ID that remains the same through the session. We can do something like the following?

```php
add_action( 'init', function() {
    if ( ! session_id() ) {
        session_start();
    }
} );
```

Few points from the docs:

1. Your application generates a unique `sessionId` when a user first visits your website or uses your application. You must use the same `sessionId` in all events throughout the session. Amazon Personalize uses the `sessionId` to associate events with the user before they log in (is anonymous).
2. To provide real-time personalization for anonymous users, specify the `sessionId` as the `userId` in your `GetRecommendations` or `GetActionRecommendations` request.